doc/manual/contracts.md

```diff
@@ -782,134 +717,200 @@ error: contract broken by the value of `fail`
 See how the command `:query config.data` succeeds, although the field `fail`
 causes an error when evaluated.
 
-A large configuration will most probably have a root contract attached,
-corresponding to its schema. If this contract checked everything eagerly,
-forcing the evaluation of most of the configuration, we would lose the benefits
-of laziness. Thus, *we want contracts to be lazy as well*. In particular, a
-subcontract attached to a field should only fire when the field is requested.
-This is the case with record contracts, and in general with all native contracts
-combinators.
+A large configuration usually has a root contract attached, corresponding to its
+schema. If this contract would perform all the checks immediately, forcing the
+evaluation of most of the configuration, we would lose the benefits of laziness.
+Thus, *we want contracts to be lazy as well*. In particular, when writing a
+contract `{foo | FooContract}`, we want that `FooContract` fires only when the
+field `foo` is requested. This is indeed the case: built-in contract
+combinators, that is array contracts, dictionary contracts and function
+contracts, all have a *delayed part*.
+
+### Immediate and delayed
+
+The most general form of contract thus has two parts:
+
+- the *immediate part* which is evaluated first and produces an answer right
+    away. This answer is either:
+  - `'Ok` to signal immediate success and forego the delayed checks (if any)
+  - `'Proceed` to signal that the immediate part didin't detect any error, but
+      the delayed checks must still be performed
+  - `'Error {..}` to signal immediate failure
+
+  In practice, the immediate part is the first argument to
+  `std.contract.custom`, and is indeed a function from an unknwon value to one
+  of the three possible answers, of type:
+
+  ```nickel
+  Dyn -> [|
+    'Ok,
+    'Proceed,
+    'Error {
+      message | String | optional,
+      notes | Array String | optional
+    }
+  |]
+  ```
+
+- the *delayed part* which returns the value with delayed checks integrated
+    inside. Those checks will fire only when further data is requested. This is
+    the second argument of `std.contract.custom`. The delayed part is a function
+    taking a label, the value, and return the augmented value, of type:
+
+  ```nickel
+  Dyn -> Dyn -> Dyn
+  ```
+
+Take the record contract `{foo | FooContract}`:
+
+- the immediate part of this contract will check that the value is a record and
+    that its only field is `foo`.
+- the delayed part will push is the `FooContract` contract (more
+    precisely map lazily) inside the value. The resulting enriched value is
+    returned.
```

Comment:
Maybe it's clearer (especially the delayed part) if you say what happens when applying `{foo | FooContract}` to the value `{foo = bar}`. The second point could be something like "the delayed part will push the `FooContract` contract inside the record, returning `{ foo = (bar | FooContract) }`"